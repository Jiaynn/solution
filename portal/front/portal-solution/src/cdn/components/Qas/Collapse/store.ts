import { observable, action, makeObservable } from 'mobx'
import autobind from 'autobind-decorator'
import Store from 'qn-fe-core/store'

export default class CollapseStore extends Store {
  @observable.ref collapseKeys: string[] = []

  constructor(initialKeys: string[]) {
    super()
    makeObservable(this)
    this.collapseKeys = initialKeys
  }

  @action.bound updateCollapseKeys(keys: string[]) {
    this.collapseKeys = keys
  }

  @autobind
  isPanelCollapsed(key: string) {
    return this.collapseKeys.indexOf(key) >= 0
  }
}
