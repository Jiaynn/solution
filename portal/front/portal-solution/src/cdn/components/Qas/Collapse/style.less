.qas-prime-panel-header {
  text-align: left;

  .panel-header-title {
    font-size: 16px;
  }

  .panel-header-collapse {
    padding-left: 10px;
    color: #2F94FA;
  }
}

.qas-prime-collapse-panel {
  padding-bottom: 32px;
}

.qas-prime-collapse {
  .ant-collapse-item {
    &:first-child {
      border-top: 1px solid #dcdfe6;
    }
  }
}