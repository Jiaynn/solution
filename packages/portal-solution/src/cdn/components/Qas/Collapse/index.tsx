/**
 * @file 折叠面板
 * @author linchen <linchen@qiniu.com>
 */

import React from 'react'
import { observer } from 'mobx-react'
import classNames from 'classnames'

import Icon from 'react-icecream/lib/icon'
import IcecreamCollapse, { CollapseProps, CollapsePanelProps } from 'react-icecream/lib/collapse'

import CollapseStore from './store'

import './style.less'

export { default as CollapseStore } from './store'

interface IPanelHeaderProps {
  title: string
  collapsed: boolean
}

export const PanelHeader = observer((props: IPanelHeaderProps) => (
  <div className="qas-prime-panel-header">
    <span className="panel-header-title">{props.title}</span>
    <span className="panel-header-collapse">
      {
        props.collapsed
        ? (<>收起<Icon type="down" /></>)
        : (<>展开<Icon type="right" /></>)
      }
    </span>
  </div>
))

export interface IPanelProps extends Omit<CollapsePanelProps, 'header'> {
  title: string
  panelKey?: string
  store: CollapseStore
}

@observer
export class Panel extends React.Component<IPanelProps> {
  renderHeader() {
    return (
      <PanelHeader
        title={this.props.title}
        collapsed={this.props.panelKey != null && this.props.store.isPanelCollapsed(this.props.panelKey)}
      />
    )
  }

  render() {
    // eslint-disable-next-line
    const { className, store, ...defaultProps } = this.props
    return (
      <IcecreamCollapse.Panel
        {...defaultProps}
        className={classNames('qas-prime-collapse-panel', className)}
        header={this.renderHeader()}
      />
    )
  }
}

export default observer(function Collapse({ className, ...defaultProps }: React.PropsWithChildren<CollapseProps>) {
  return (
    <IcecreamCollapse
      {...defaultProps}
      bordered={false}
      className={classNames('qas-prime-collapse', className)}
      expandIcon={() => null}
    />
  )
})
