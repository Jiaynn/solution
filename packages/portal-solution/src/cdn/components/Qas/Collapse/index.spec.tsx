/*
 * @file cases for collapse component
 * @author linchen <linchen@qiniu.com>
 */

import React from 'react'

import { RendererUtils as Renderer } from 'test'
import Collapse, { Panel, CollapseStore } from '.'

it('renders correctly without panel', () => {
  const tree = new Renderer().createWithAct(
    <Collapse />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with panel', () => {
  const store = new CollapseStore(['panel1', 'panel2'])
  const tree = new Renderer().createWithAct(
    <Collapse>
      <Panel title="test panel 1" key="panel1" store={store} />
      <Panel title="test panel 2" key="panel2" store={store} />
    </Collapse>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
