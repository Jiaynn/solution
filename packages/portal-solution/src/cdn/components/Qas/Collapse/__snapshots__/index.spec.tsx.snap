// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with panel 1`] = `
<div
  className="ant-collapse ant-collapse-borderless ant-collapse-icon-position-left qas-prime-collapse"
  role={null}
>
  <div
    className="ant-collapse-item qas-prime-collapse-panel"
  >
    <div
      aria-expanded="false"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="qas-prime-panel-header"
      >
        <span
          className="panel-header-title"
        >
          test panel 1
        </span>
        <span
          className="panel-header-collapse"
        >
          收起
          <i
            aria-label="icon: down"
            className="anticon anticon-down"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-collapse-item qas-prime-collapse-panel"
  >
    <div
      aria-expanded="false"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="qas-prime-panel-header"
      >
        <span
          className="panel-header-title"
        >
          test panel 2
        </span>
        <span
          className="panel-header-collapse"
        >
          收起
          <i
            aria-label="icon: down"
            className="anticon anticon-down"
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly without panel 1`] = `
<div
  className="ant-collapse ant-collapse-borderless ant-collapse-icon-position-left qas-prime-collapse"
  role={null}
/>
`;
